File: Java/src/session2/td/heritageMultiple/NombreDecimal.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public abstract class NombreDecimal implements Nat {

    public String chiffres;

    public NombreDecimal(String chiffres) {
        this.chiffres = chiffres;
    }

    @Override
    public boolean estNul() {
        // Peut être simplifié du fait du nettoyage ("0" pour zéro)
        for(int i = 0; i < this.taille(); i++){
            if(this.chiffre(i) != 0){
                return false;
            }
        }
        return true;
    }

    @Override
    public Nat predecesseur() {
        if(this.estNul()){
            throw new UnsupportedOperationException();
        }
        int t = this.taille();
        StringBuilder rep = new StringBuilder();
        int retenue = -1;
        for(int i = 0; i < t; i++){
            int chiffre = this.chiffre(i) + retenue;
            if(chiffre == -1){
                chiffre = 9;
                retenue = -1;
            }else{
                retenue = 0;
            }
            rep.append(Integer.toString(chiffre));
        }
        return this.creerNatAvecRepresentation(rep.reverse().toString());
    }
    @Override
    public int chiffre(int i){
        if(i < this.taille())
            return Character.getNumericValue(chiffres.charAt(chiffres.length() -1 -i));
        return 0;
    }

    @Override
    public int taille() {
        return this.chiffres.length();
    }

    @Override
    public int val() {
        return Integer.parseInt(this.chiffres);
    }

    @Override
    public Nat creerNatAvecValeur(int val) {
        return creerNatAvecRepresentation(Integer.toString(val));
    }

    @Override
    public Nat creerZero() {
        return this.creerNatAvecValeur(0);
    }

    @Override
    public Nat creerSuccesseur(Nat predecesseur) {
        int t = predecesseur.taille();
        StringBuilder rep = new StringBuilder();
        int retenue = 1;
        for(int i = 0; i < t; i++){
            int chiffre = predecesseur.chiffre(i) + retenue;
            if(chiffre > 9){
                chiffre = chiffre - 10;
                retenue = 1;
            }else{
                retenue = 0;
            }
            rep.append(Integer.toString(chiffre));
        }
        rep = retenue == 0 ? rep : rep.append(1);
        return this.creerNatAvecRepresentation(rep.reverse().toString());
    }

}

File: TypeScript/src/session2.td/heritageMultiple/NombreDecimal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NombreDecimal = /** @class */ (function () {
    function NombreDecimal(chiffres) {
        this._chiffres = chiffres;
    }
    NombreDecimal.prototype.estNul = function () {
        // Peut être simplifié du fait du nettoyage ("0" pour zéro)
        for (var i = 0; i < this.taille(); i++) {
            if (this.chiffre(i) != 0) {
                return false;
            }
        }
        return true;
    };
    NombreDecimal.prototype.predecesseur = function () {
        if (this.estNul()) {
            throw new Error();
        }
        var t = this.taille();
        var rep = "";
        var retenue = -1;
        for (var i = 0; i < t; i++) {
            var chiffre = this.chiffre(i) + retenue;
            if (chiffre == -1) {
                chiffre = 9;
                retenue = -1;
            }
            else {
                retenue = 0;
            }
            rep += +chiffre;
        }
        return this.creerNatAvecRepresentation(rep.split('').reverse().join(''));
    };
    NombreDecimal.prototype.chiffre = function (i) {
        if (i < this.taille())
            return +(this._chiffres.charAt(this._chiffres.length - 1 - i));
        return 0;
    };
    NombreDecimal.prototype.taille = function () {
        return this._chiffres.length;
    };
    NombreDecimal.prototype.val = function () {
        return +this._chiffres;
    };
    NombreDecimal.prototype.creerNatAvecValeur = function (val) {
        return this.creerNatAvecRepresentation(String(val));
    };
    NombreDecimal.prototype.creerZero = function () {
        return this.creerNatAvecValeur(0);
    };
    NombreDecimal.prototype.creerSuccesseur = function (predecesseur) {
        var t = predecesseur.taille();
        var rep = "";
        var retenue = 1;
        for (var i = 0; i < t; i++) {
            var chiffre = predecesseur.chiffre(i) + retenue;
            if (chiffre > 9) {
                chiffre = chiffre - 10;
                retenue = 1;
            }
            else {
                retenue = 0;
            }
            rep += +chiffre;
        }
        rep = retenue == 0 ? rep : rep += +(1);
        return this.creerNatAvecRepresentation(rep.split('').reverse().join(''));
    };
    return NombreDecimal;
}());
exports.NombreDecimal = NombreDecimal;
